### PR TITLE
implement support for encoding name constraints

### PR DESCRIPTION
**Pull request by reaperhulk**
refs #2192

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 1]
```diff
@@ -58,6 +58,42 @@
 _MemoryBIO = collections.namedtuple("_MemoryBIO", ["bio", "char_ptr"])
 
 
+def _encode_name_constraints(backend, name_constraints):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Can these be organized somewhere down bleow, with the rest of the extensions

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Sure. (This module is pretty unwieldy so we should probably move this stuff to another module, but that should be in another PR)

File: CHANGELOG.rst [hunk 2]
```diff
@@ -9,6 +9,8 @@ Changelog
 * The :class:`~cryptography.x509.Certificate` class now has
   :attr:`~cryptography.x509.Certificate.signature` and
   :attr:`~cryptography.x509.Certificate.tbs_certificate_bytes` attributes.
+* :class:`~cryptography.x509.NameConstraints` are now supported in the
+  :class:`~cryptography.x509.CertificateBuilder`.
```
Comment (alex) [inline CHANGELOG.rst]: CSR builder as well, I guess that's probably not practically relevant though?

Comment (reaperhulk) [inline CHANGELOG.rst]: Some users might care (sometimes internal CAs allow issuance using extensions requested in the CSR), so I'll go ahead and mention CSRBuilder supports it too.

**Comment by sigmavirus24** [comment]
I know nothing @mention-bot =P

**Comment by reaperhulk** [comment]
Don't be so modest, it's touching code you had a hand in Ian! :grinning:
